spring-rabbit/src/main/java/org/springframework/amqp/rabbit/annotation/MultiRabbitBootstrapConfiguration.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2002-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.amqp.rabbit.annotation;
+
+import org.springframework.amqp.rabbit.config.RabbitListenerConfigUtils;
+import org.springframework.amqp.rabbit.listener.RabbitListenerEndpointRegistry;
+import org.springframework.beans.factory.support.BeanDefinitionRegistry;
+import org.springframework.beans.factory.support.RootBeanDefinition;
+import org.springframework.context.annotation.ImportBeanDefinitionRegistrar;
+import org.springframework.core.type.AnnotationMetadata;
+
+/**
+ * An {@link ImportBeanDefinitionRegistrar} class that registers a
+ * {@link MultiRabbitListenerAnnotationBeanPostProcessor} (overwriting the regular
+ * {@link RabbitListenerAnnotationBeanPostProcessor}) bean capable of processing
+ * Spring's {@link RabbitListener} annotations.
+ * <p>
+ * This configuration must be imported manually.
+ *
+ * @author Wander Costa
+ * @see MultiRabbitListenerAnnotationBeanPostProcessor
+ * @see RabbitListenerAnnotationBeanPostProcessor
+ * @see RabbitListenerEndpointRegistry
+ * @see EnableRabbit
+ */
+public class MultiRabbitBootstrapConfiguration implements ImportBeanDefinitionRegistrar {
```

Comment:
Don't we need an annotation or `DeferredImportSelector` to trigger this to ensure it's loaded before the `RabbitBootstrapConfiguration`?

cc/ @artembilan

Reply:
Can't the user just declare a bean with the subclass and `RABBIT_LISTENER_ANNOTATION_PROCESSOR_BEAN_NAME` since the `RabbitListenerConfigurationSelector` won't run until all `@Configuration` classes have been processed; removing the need for this class altogether.

Reply:
Well, need to understand how it is tested over here though.
Probably just `EnableRabbit(multi = true)` would be enough...

Reply:
I'm not too sure how to proceed here. No changes.

Reply:
I think we can just delete this; in order to override the default BPP, simply define the subclass as a bean with the default name, and the default one will be skipped here https://github.com/spring-projects/spring-amqp/blob/715f39ff3baab340e5d681d74697140c88a7874e/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/annotation/RabbitBootstrapConfiguration.java#L51-L56

Reply:
Hi @garyrussell & @artembilan,
I'm writing this here to connect to the context of the issue.

I'm working on the boot part, and I tried to implement the instantiation of MultRabbitBPP the way suggested. But as overriding of beans is now disabled by default in Spring, it fails to register a bean with the same name as `RabbitListenerConfigUtils.RABBIT_LISTENER_ANNOTATION_PROCESSOR_BEAN_NAME`.

I'm trying to work around it by providing a MultiRabbit BPP bean with another name, via configuration. But when processing listeners, it still picks up the original one BPP and ALSO tries to process MultiRabbitBPP, leading to an error of multiple registrations of the same declarables.

Maybe it should never instantiate `RabbitListenerAnnotationBeanPostProcessor` since `MultiRabbitListenerAnnotationBeanPostProcessor` is provided, but I'm not figuring out how to make it work, or if that's related to some order of processing.

I got to make this work by directly implementing `RabbitBootstrapConfiguration` to instantiate `new RootBeanDefinition(MultiRabbitListenerAnnotationBeanPostProcessor.class)` whenever `EnableRabbit(multi = true)` (as per @artembilan's [suggestion](https://github.com/spring-projects/spring-amqp/pull/1111#discussion_r442497506)), but I feel this is not the best solution (very intrusive and not SOLID).

Would you have any suggestions on how to proceed with this?

Reply:
@rwanderc I am not sure what you mean by SOLID but another, perhaps less intrusive, option would be to add a new Boot property `multiCluster`. While we can't reference the `RabbitProperties` we can get a reference to the environment:

```java
public class RabbitBootstrapConfiguration implements ImportBeanDefinitionRegistrar, EnvironmentAware {

	private Environment environment;

	@Override
	public void registerBeanDefinitions(@Nullable AnnotationMetadata importingClassMetadata,
			BeanDefinitionRegistry registry) {

		String multiS =  this.environment.getProperty("spring.rabbitmq.multi-cluster");
		boolean multi = multiS != null && Boolean.parseBoolean(multiS);
		if (multi) {
			...
		}
		else if (!registry.containsBeanDefinition(
				RabbitListenerConfigUtils.RABBIT_LISTENER_ANNOTATION_PROCESSOR_BEAN_NAME)) {

			registry.registerBeanDefinition(RabbitListenerConfigUtils.RABBIT_LISTENER_ANNOTATION_PROCESSOR_BEAN_NAME,
					new RootBeanDefinition(RabbitListenerAnnotationBeanPostProcessor.class));
		}

		if (!registry.containsBeanDefinition(RabbitListenerConfigUtils.RABBIT_LISTENER_ENDPOINT_REGISTRY_BEAN_NAME)) {
			registry.registerBeanDefinition(RabbitListenerConfigUtils.RABBIT_LISTENER_ENDPOINT_REGISTRY_BEAN_NAME,
					new RootBeanDefinition(RabbitListenerEndpointRegistry.class));
		}
	}

	@Override
	public void setEnvironment(Environment environment) {
		this.environment = environment;
	}

}
```